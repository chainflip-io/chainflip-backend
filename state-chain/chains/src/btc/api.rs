pub mod batch_transfer;

use super::{
	deposit_address::DepositAddress, AggKey, Bitcoin, BitcoinCrypto, BitcoinOutput, BtcAmount,
	Utxo, BITCOIN_DUST_LIMIT, CHANGE_ADDRESS_SALT,
};
use crate::*;
use frame_support::{CloneNoBound, DebugNoBound, EqNoBound, Never, PartialEqNoBound};
use sp_std::marker::PhantomData;

#[derive(CloneNoBound, DebugNoBound, PartialEqNoBound, EqNoBound, Encode, Decode, TypeInfo)]
#[scale_info(skip_type_params(Environment))]
pub enum BitcoinApi<Environment: 'static> {
	BatchTransfer(batch_transfer::BatchTransfer),
	#[doc(hidden)]
	#[codec(skip)]
	_Phantom(PhantomData<Environment>, Never),
}

pub type SelectedUtxosAndChangeAmount = (Vec<Utxo>, BtcAmount);

#[derive(Copy, Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
pub enum UtxoSelectionType {
	SelectAllForRotation,
	SelectForConsolidation,
	Some { output_amount: BtcAmount, number_of_outputs: u64 },
}

impl<E> ConsolidateCall<Bitcoin> for BitcoinApi<E>
where
	E: ChainEnvironment<UtxoSelectionType, SelectedUtxosAndChangeAmount>
		+ ChainEnvironment<(), AggKey>,
{
	fn consolidate_utxos() -> Result<Self, ConsolidationError> {
		let agg_key @ AggKey { current, .. } =
			<E as ChainEnvironment<(), AggKey>>::lookup(()).ok_or(ConsolidationError::Other)?;
		let bitcoin_change_script =
			DepositAddress::new(current, CHANGE_ADDRESS_SALT).script_pubkey();

		let (selected_input_utxos, change_amount) =
			E::lookup(UtxoSelectionType::SelectForConsolidation)
				.ok_or(ConsolidationError::NotRequired)?;

		log::info!("Consolidating {} btc utxos", selected_input_utxos.len());

		let btc_outputs =
			vec![BitcoinOutput { amount: change_amount, script_pubkey: bitcoin_change_script }];

		Ok(Self::BatchTransfer(batch_transfer::BatchTransfer::new_unsigned(
			&agg_key,
			agg_key.current,
			selected_input_utxos,
			btc_outputs,
		)))
	}
}

impl<E> AllBatch<Bitcoin> for BitcoinApi<E>
where
	E: ChainEnvironment<UtxoSelectionType, SelectedUtxosAndChangeAmount>
		+ ChainEnvironment<(), AggKey>,
{
	fn new_unsigned(
		_fetch_params: Vec<FetchAssetParams<Bitcoin>>,
		transfer_params: Vec<TransferAssetParams<Bitcoin>>,
	) -> Result<Self, AllBatchError> {
		let agg_key @ AggKey { current, .. } =
			<E as ChainEnvironment<(), AggKey>>::lookup(()).ok_or(AllBatchError::Other)?;
		let bitcoin_change_script =
			DepositAddress::new(current, CHANGE_ADDRESS_SALT).script_pubkey();
		let mut total_output_amount: u64 = 0;
		let mut btc_outputs = vec![];
		for transfer_param in transfer_params {
			if transfer_param.amount >= BITCOIN_DUST_LIMIT {
				btc_outputs.push(BitcoinOutput {
					amount: transfer_param.amount,
					script_pubkey: transfer_param.to,
				});
				total_output_amount += transfer_param.amount;
			}
		}
		// Looks up all available Utxos and selects and takes them for the transaction depending on
		// the amount that needs to be output. If the output amount is 0,
		let (selected_input_utxos, change_amount) = E::lookup(UtxoSelectionType::Some {
			output_amount: (total_output_amount > 0)
				.then_some(total_output_amount)
				.ok_or(AllBatchError::NotRequired)?,
			number_of_outputs: (btc_outputs.len() + 1) as u64, // +1 for the change output
		})
		.ok_or(AllBatchError::Other)?;
		if change_amount >= BITCOIN_DUST_LIMIT {
			btc_outputs.push(BitcoinOutput {
				amount: change_amount,
				script_pubkey: bitcoin_change_script,
			});
		}

		Ok(Self::BatchTransfer(batch_transfer::BatchTransfer::new_unsigned(
			&agg_key,
			agg_key.current,
			selected_input_utxos,
			btc_outputs,
		)))
	}
}

impl<E> SetAggKeyWithAggKey<BitcoinCrypto> for BitcoinApi<E>
where
	E: ChainEnvironment<UtxoSelectionType, SelectedUtxosAndChangeAmount>,
{
	fn new_unsigned(
		maybe_old_key: Option<<BitcoinCrypto as ChainCrypto>::AggKey>,
		new_key: <BitcoinCrypto as ChainCrypto>::AggKey,
	) -> Result<Self, SetAggKeyWithAggKeyError> {
		// We will use the bitcoin address derived with the salt of 0 as the vault address where we
		// collect unspent amounts in btc transactions and consolidate funds when rotating epoch.
		let new_vault_change_script =
			DepositAddress::new(new_key.current, CHANGE_ADDRESS_SALT).script_pubkey();

		// Max possible btc value to get all available utxos
		// If we don't have any UTXOs then we're not required to do this.
		let (all_input_utxos, change_amount) = E::lookup(UtxoSelectionType::SelectAllForRotation)
			.ok_or(SetAggKeyWithAggKeyError::NotRequired)?;

		Ok(Self::BatchTransfer(batch_transfer::BatchTransfer::new_unsigned(
			&maybe_old_key.ok_or(SetAggKeyWithAggKeyError::Failed)?,
			new_key.current,
			all_input_utxos,
			vec![BitcoinOutput { amount: change_amount, script_pubkey: new_vault_change_script }],
		)))
	}
}

impl<E> From<batch_transfer::BatchTransfer> for BitcoinApi<E> {
	fn from(tx: batch_transfer::BatchTransfer) -> Self {
		Self::BatchTransfer(tx)
	}
}

// TODO: Implement transfer / transfer and call for Bitcoin.
impl<E: ReplayProtectionProvider<Bitcoin>> ExecutexSwapAndCall<Bitcoin> for BitcoinApi<E> {
	fn new_unsigned(
		_transfer_param: TransferAssetParams<Bitcoin>,
		_source_chain: ForeignChain,
		_source_address: Option<ForeignChainAddress>,
		_gas_budget: <Bitcoin as Chain>::ChainAmount,
		_message: Vec<u8>,
	) -> Result<Self, DispatchError> {
		Err(DispatchError::Other("Bitcoin's ExecutexSwapAndCall is not supported."))
	}
}

// transfer_fallback is unsupported for Bitcoin.
impl<E: ReplayProtectionProvider<Bitcoin>> TransferFallback<Bitcoin> for BitcoinApi<E> {
	fn new_unsigned(_transfer_param: TransferAssetParams<Bitcoin>) -> Result<Self, DispatchError> {
		Err(DispatchError::Other("Bitcoin's TransferFallback is not supported."))
	}
}

impl<E> ApiCall<BitcoinCrypto> for BitcoinApi<E> {
	fn threshold_signature_payload(&self) -> <BitcoinCrypto as ChainCrypto>::Payload {
		match self {
			BitcoinApi::BatchTransfer(tx) => tx.threshold_signature_payload(),

			BitcoinApi::_Phantom(..) => unreachable!(),
		}
	}

	fn signed(
		self,
		threshold_signature: &<BitcoinCrypto as ChainCrypto>::ThresholdSignature,
	) -> Self {
		match self {
			BitcoinApi::BatchTransfer(call) => call.signed(threshold_signature).into(),

			BitcoinApi::_Phantom(..) => unreachable!(),
		}
	}

	fn chain_encoded(&self) -> Vec<u8> {
		match self {
			BitcoinApi::BatchTransfer(call) => call.chain_encoded(),

			BitcoinApi::_Phantom(..) => unreachable!(),
		}
	}

	fn is_signed(&self) -> bool {
		match self {
			BitcoinApi::BatchTransfer(call) => call.is_signed(),

			BitcoinApi::_Phantom(..) => unreachable!(),
		}
	}

	fn transaction_out_id(&self) -> <BitcoinCrypto as ChainCrypto>::TransactionOutId {
		match self {
			BitcoinApi::BatchTransfer(call) => call.transaction_out_id(),
			BitcoinApi::_Phantom(..) => unreachable!(),
		}
	}
}
