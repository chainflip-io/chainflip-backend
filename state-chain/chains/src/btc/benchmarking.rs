#![cfg(feature = "runtime-benchmarks")]

use sp_std::{vec, vec::Vec};

use crate::benchmarking_value::{BenchmarkValue, BenchmarkValueExtended};

use super::{
	api::{batch_transfer::BatchTransfer, BitcoinApi},
	deposit_address::DepositAddress,
	AggKey, BitcoinFeeInfo, BitcoinFetchId, BitcoinOutput, BitcoinTrackedData,
	BitcoinTransactionData, PreviousOrCurrent, ScriptPubkey, Signature, SigningPayload, Utxo,
	UtxoId,
};

impl BenchmarkValue for AggKey {
	fn benchmark_value() -> Self {
		AggKey { previous: None, current: [2u8; 32] }
	}
}

impl BenchmarkValue for BitcoinTransactionData {
	fn benchmark_value() -> Self {
		Self { encoded_transaction: vec![1u8; 100] }
	}
}

impl<T: BenchmarkValue> BenchmarkValue for Vec<T> {
	fn benchmark_value() -> Self {
		vec![T::benchmark_value()]
	}
}

impl BenchmarkValue for UtxoId {
	fn benchmark_value() -> Self {
		UtxoId { tx_id: [1u8; 32].into(), vout: 1 }
	}
}

// Bitcoin threshold signature
impl BenchmarkValue for Signature {
	fn benchmark_value() -> Self {
		[0xau8; 64]
	}
}

// Bitcoin payload
impl BenchmarkValue for SigningPayload {
	fn benchmark_value() -> Self {
		[1u8; 32]
	}
}

impl BenchmarkValue for BitcoinFetchId {
	fn benchmark_value() -> Self {
		Self(1)
	}
}

impl BenchmarkValueExtended for BitcoinFetchId {
	fn benchmark_value_by_id(id: u8) -> Self {
		Self(id.into())
	}
}

impl<E> BenchmarkValue for BitcoinApi<E> {
	fn benchmark_value() -> Self {
		BitcoinApi::BatchTransfer(BatchTransfer::new_unsigned(
			&BenchmarkValue::benchmark_value(),
			BenchmarkValue::benchmark_value(),
			vec![Utxo {
				amount: Default::default(),
				id: BenchmarkValue::benchmark_value(),
				deposit_address: DepositAddress::new(Default::default(), Default::default()),
			}],
			vec![BitcoinOutput {
				amount: Default::default(),
				script_pubkey: BenchmarkValue::benchmark_value(),
			}],
		))
	}
}

impl BenchmarkValue for BitcoinTrackedData {
	fn benchmark_value() -> Self {
		BitcoinTrackedData { btc_fee_info: BitcoinFeeInfo::new(4321) }
	}
}

impl BenchmarkValue for PreviousOrCurrent {
	fn benchmark_value() -> Self {
		Self::Current
	}
}

impl BenchmarkValue for ScriptPubkey {
	fn benchmark_value() -> Self {
		Self::benchmark_value_by_id(0)
	}
}

impl BenchmarkValueExtended for ScriptPubkey {
	fn benchmark_value_by_id(id: u8) -> Self {
		Self::Taproot([id; 32])
	}
}
