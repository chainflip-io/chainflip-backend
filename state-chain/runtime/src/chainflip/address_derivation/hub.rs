// Copyright 2025 Chainflip Labs GmbH
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

use crate::Environment;
use cf_chains::{
	address::{AddressDerivationApi, AddressDerivationError},
	hub::calculate_derived_address,
	Assethub, Chain,
};
use cf_primitives::ChannelId;

use super::AddressDerivation;

impl AddressDerivationApi<Assethub> for AddressDerivation {
	fn generate_address(
		_source_asset: <Assethub as Chain>::ChainAsset,
		channel_id: ChannelId,
	) -> Result<<Assethub as Chain>::ChainAccount, AddressDerivationError> {
		let master_account = Environment::assethub_vault_account()
			.ok_or(AddressDerivationError::MissingAssethubVault)?;

		Ok(calculate_derived_address(master_account, channel_id))
	}

	fn generate_address_and_state(
		source_asset: <Assethub as Chain>::ChainAsset,
		channel_id: ChannelId,
	) -> Result<
		(<Assethub as Chain>::ChainAccount, <Assethub as Chain>::DepositChannelState),
		AddressDerivationError,
	> {
		Ok((
			<Self as AddressDerivationApi<Assethub>>::generate_address(source_asset, channel_id)?,
			Default::default(),
		))
	}
}

#[cfg(test)]
mod test {
	use super::*;
	use crate::Runtime;
	use cf_chains::dot::PolkadotAccountId;
	use cf_primitives::chains::assets::hub;
	use frame_support::sp_runtime::app_crypto::Ss58Codec;
	use pallet_cf_environment::AssethubVaultAccountId;

	#[test]
	fn single_layer() {
		sp_io::TestExternalities::new_empty().execute_with(|| {
			let (account_id, address_format) =
				sp_runtime::AccountId32::from_ss58check_with_version(
					"15uPkKV7SsNXxw5VCu3LgnuaR5uSZ4QMyzxnLfDFE9J5nni9",
				)
				.unwrap();
			AssethubVaultAccountId::<Runtime>::put(PolkadotAccountId::from_aliased(
				*account_id.as_ref(),
			));

			assert_eq!(
				"12AeXofJkQErqQuiJmJapqwS4KiAZXBhAYoj9HZ2sYo36mRg",
				sp_runtime::AccountId32::new(
					*<AddressDerivation as AddressDerivationApi<Assethub>>::generate_address(
						hub::Asset::HubUsdc,
						6259
					)
					.unwrap()
					.aliased_ref()
				)
				.to_ss58check_with_version(address_format),
			);
		});
	}

	#[test]
	fn four_layers() {
		sp_io::TestExternalities::new_empty().execute_with(|| {
			let (alice, address_format) = sp_runtime::AccountId32::from_ss58check_with_version(
				"15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5",
			)
			.unwrap();
			AssethubVaultAccountId::<Runtime>::put(PolkadotAccountId::from_aliased(
				*alice.as_ref(),
			));

			assert_eq!(
				// The following account was generated using nested utility.asDerivative calls in
				// PolkaJS. The wrapped call was system.remarkWithEvent, which emits the generated
				// address in its event.
				//
				// The call was dispatched from the Alice account.
				//
				// To see the call go to
				// extrinsics/decode/0x1a0101001a0102001a0103001a01040000071448414c4c4f
				// on any polkaJS instance connected to a polkadot node.
				//
				// Call details:
				// 1a01 is the utility.asDerivative call index.
				// 0007 is the system.remarkWithEvent call index.
				// Encoded call: 0x1a01 0100 1a01 0200 1a01 0300 1a01 0400 0007 1448414c4c4f
				//                 ---- -^-- ---- -^-- ---- -^-- ---- -^--      b"HALLO"
				"1422Jc2BYRh5ENjxWJchoHPSC2Rd4jFs8PDWHqBJue4yskEt",
				sp_runtime::AccountId32::new(
					*<AddressDerivation as AddressDerivationApi<Assethub>>::generate_address(
						hub::Asset::HubDot,
						0x0004_0003_0002_0001
					)
					.unwrap()
					.aliased_ref()
				)
				.to_ss58check_with_version(address_format),
			);
		});
	}
}
