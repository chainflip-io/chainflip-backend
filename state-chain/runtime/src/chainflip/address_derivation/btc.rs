use super::AddressDerivation;
use crate::BitcoinVault;
use cf_chains::{
	address::{AddressDerivationApi, AddressDerivationError},
	btc::deposit_address::DepositAddress,
	Bitcoin, Chain,
};
use cf_primitives::ChannelId;
use cf_traits::KeyProvider;

impl AddressDerivationApi<Bitcoin> for AddressDerivation {
	fn generate_address(
		source_asset: <Bitcoin as Chain>::ChainAsset,
		channel_id: ChannelId,
	) -> Result<<Bitcoin as Chain>::ChainAccount, AddressDerivationError> {
		<Self as AddressDerivationApi<Bitcoin>>::generate_address_and_state(
			source_asset,
			channel_id,
		)
		.map(|(address, _)| address)
	}

	fn generate_address_and_state(
		_source_asset: <Bitcoin as Chain>::ChainAsset,
		channel_id: ChannelId,
	) -> Result<
		(<Bitcoin as Chain>::ChainAccount, <Bitcoin as Chain>::DepositChannelState),
		AddressDerivationError,
	> {
		let channel_id: u32 = channel_id
			.try_into()
			.map_err(|_| AddressDerivationError::BitcoinChannelIdTooLarge)?;

		let channel_state = DepositAddress::new(
			// TODO: The key should be passed as an argument (or maybe KeyProvider type arg).
			BitcoinVault::active_epoch_key()
				.ok_or(AddressDerivationError::MissingBitcoinVault)?
				.key
				.current,
			channel_id,
		);

		Ok((channel_state.script_pubkey(), channel_state))
	}
}

#[test]
fn test_address_generation() {
	use crate::Runtime;
	use cf_chains::Bitcoin;
	use cf_primitives::chains::assets::btc;
	use cf_utilities::assert_ok;
	use pallet_cf_validator::CurrentEpoch;
	use pallet_cf_vaults::{CurrentVaultEpoch, Vault, Vaults};

	sp_io::TestExternalities::new_empty().execute_with(|| {
		CurrentEpoch::<Runtime>::set(1);
		Vaults::<Runtime, crate::BitcoinInstance>::insert(
			1,
			Vault::<Bitcoin> {
				public_key: cf_chains::btc::AggKey {
					previous: None,
					current: hex_literal::hex!(
						"9fe94d03955ff4cc5dec97fa5f0dc564ae5ab63012e76dbe84c87c1c83460b48"
					),
				},
				active_from_block: 1,
			},
		);
		CurrentVaultEpoch::<Runtime, crate::BitcoinInstance>::put(1);
		assert_ok!(<AddressDerivation as AddressDerivationApi<Bitcoin>>::generate_address(
			btc::Asset::Btc,
			1
		));
	});
}
