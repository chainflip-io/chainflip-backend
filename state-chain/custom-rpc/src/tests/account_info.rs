use super::*;

#[test]
fn test_no_account_serialization() {
	let wrapper = RpcAccountInfoWrapper {
		common_items: RpcAccountInfoCommonItems {
			flip_balance: 1500000000000000000u128.into(), // 1.5 FLIP
			asset_balances: any::AssetMap {
				eth: eth::AssetMap {
					eth: 2500000000000000000u128.into(), // 2.5 ETH
					flip: 0u128.into(),
					usdc: 1000000u128.into(), // 1 USDC
					usdt: 0u128.into(),
				},
				btc: btc::AssetMap { btc: 100000000u128.into() }, // 1 BTC
				dot: dot::AssetMap { dot: 10000000000u128.into() }, // 1 DOT
				arb: arb::AssetMap { eth: 0u128.into(), usdc: 0u128.into() },
				sol: sol::AssetMap { sol: 0u128.into(), usdc: 0u128.into() },
				hub: hub::AssetMap { dot: 0u128.into(), usdc: 0u128.into(), usdt: 0u128.into() },
			},
			bond: 0u32.into(),
			estimated_redeemable_balance: 0u32.into(),
			bound_redeem_address: None,
			restricted_balances: BTreeMap::new(),
			delegating_to: None,
		},
		role_specific: RpcAccountInfo::Unregistered {},
	};
	insta::assert_snapshot!(serde_json::to_string_pretty(&wrapper).unwrap());
}

#[test]
fn test_broker_serialization() {
	use cf_chains::btc::BitcoinNetwork;

	let earned_fees = any::AssetMap {
		eth: eth::AssetMap {
			eth: 250000000000000000u128.into(),   // 0.25 ETH
			flip: 1000000000000000000u128.into(), // 1 FLIP
			usdc: 500000u128.into(),              // 0.5 USDC
			usdt: 300000u128.into(),              // 0.3 USDT
		},
		btc: btc::AssetMap { btc: 5000000u128.into() }, // 0.05 BTC
		dot: dot::AssetMap { dot: 5000000000u128.into() }, // 0.5 DOT
		arb: arb::AssetMap {
			eth: 100000000000000000u128.into(), // 0.1 ETH
			usdc: 200000u128.into(),            // 0.2 USDC
		},
		sol: sol::AssetMap {
			sol: 1000000000u128.into(), // 1 SOL
			usdc: 150000u128.into(),    // 0.15 USDC
		},
		hub: hub::AssetMap {
			dot: 2000000000u128.into(), // 0.2 DOT
			usdc: 100000u128.into(),    // 0.1 USDC
			usdt: 50000u128.into(),     // 0.05 USDT
		},
	};

	let wrapper = RpcAccountInfoWrapper {
		common_items: RpcAccountInfoCommonItems {
			flip_balance: 5000000000000000000u128.into(), // 5 FLIP
			asset_balances: Default::default(),
			bond: 10000000000000000000u128.into(), // 10 FLIP
			estimated_redeemable_balance: 4000000000000000000u128.into(),
			bound_redeem_address: Some(H160::from([0xaa; 20])),
			restricted_balances: BTreeMap::from_iter(vec![
				(H160::from([0xbb; 20]), 2000000000000000000u128.into()), // 2 FLIP restricted
			]),
			delegating_to: None,
		},
		role_specific: RpcAccountInfo::Broker {
			earned_fees,
			affiliates: vec![
				RpcAffiliate {
					account_id: AccountId32::new([1; 32]),
					details: AffiliateDetails {
						short_id: 1u8.into(),
						withdrawal_address: H160::from([0xcf; 20]),
					},
				},
				RpcAffiliate {
					account_id: AccountId32::new([2; 32]),
					details: AffiliateDetails {
						short_id: 42u8.into(),
						withdrawal_address: H160::from([0xde; 20]),
					},
				},
			],
			btc_vault_deposit_address: Some(
				ScriptPubkey::Taproot([1u8; 32]).to_address(&BitcoinNetwork::Testnet),
			),
		},
	};
	insta::assert_snapshot!(serde_json::to_string_pretty(&wrapper).unwrap());
}

#[test]
fn test_lp_serialization() {
	use cf_chains::hub::SubstrateNetworkAddress;
	let refund_addresses = HashMap::from_iter(vec![
		(
			ForeignChain::Ethereum,
			Some(ForeignChainAddressHumanreadable::Eth(H160::from([0x11; 20]))),
		),
		(
			ForeignChain::Polkadot,
			Some(ForeignChainAddressHumanreadable::Dot(SubstrateNetworkAddress::polkadot(
				AccountId32::new([0xcf; 32]),
			))),
		),
		(ForeignChain::Bitcoin, None),
		(
			ForeignChain::Arbitrum,
			Some(ForeignChainAddressHumanreadable::Arb(H160::from([0x22; 20]))),
		),
		(
			ForeignChain::Solana,
			Some(ForeignChainAddressHumanreadable::Sol(
				"11111111111111111111111111111111".to_string(),
			)),
		),
	]);

	let earned_fees = any::AssetMap {
		eth: eth::AssetMap {
			eth: 1000000000000000000u128.into(), // 1 ETH
			flip: u64::MAX.into(),
			usdc: (u64::MAX / 2 - 1).into(),
			usdt: 250000u128.into(), // 0.25 USDT
		},
		btc: btc::AssetMap { btc: 25000000u128.into() }, // 0.25 BTC
		dot: dot::AssetMap { dot: 15000000000u128.into() }, // 1.5 DOT
		arb: arb::AssetMap { eth: 1u128.into(), usdc: 2u128.into() },
		sol: sol::AssetMap { sol: 2u128.into(), usdc: 4u128.into() },
		hub: hub::AssetMap {
			dot: 5000000000u128.into(),
			usdc: 1000000u128.into(),
			usdt: 500000u128.into(),
		},
	};

	let boost_balances = any::AssetMap {
		btc: btc::AssetMap {
			btc: vec![
				RpcLiquidityProviderBoostPoolInfo {
					fee_tier: 5,
					total_balance: 100_000_000u128.into(),
					available_balance: 50_000_000u128.into(),
					in_use_balance: 50_000_000u128.into(),
					is_withdrawing: false,
				},
				RpcLiquidityProviderBoostPoolInfo {
					fee_tier: 10,
					total_balance: 200_000_000u128.into(),
					available_balance: 150_000_000u128.into(),
					in_use_balance: 50_000_000u128.into(),
					is_withdrawing: true,
				},
			],
		},
		..Default::default()
	};

	let asset_balances = any::AssetMap {
		eth: eth::AssetMap {
			eth: u128::MAX.into(),
			flip: (u128::MAX / 2).into(),
			usdc: 10000000u128.into(), // 10 USDC
			usdt: 5000000u128.into(),  // 5 USDT
		},
		btc: btc::AssetMap { btc: 500000000u128.into() }, // 5 BTC
		dot: dot::AssetMap { dot: 100000000000u128.into() }, // 10 DOT
		arb: arb::AssetMap { eth: 1u128.into(), usdc: 2u128.into() },
		sol: sol::AssetMap { sol: 3u128.into(), usdc: 4u128.into() },
		hub: hub::AssetMap {
			dot: 50000000000u128.into(),
			usdc: 3000000u128.into(),
			usdt: 2000000u128.into(),
		},
	};

	let wrapper = RpcAccountInfoWrapper {
		common_items: RpcAccountInfoCommonItems {
			flip_balance: 25000000000000000000u128.into(), // 25 FLIP
			asset_balances,
			bond: 0u32.into(),
			estimated_redeemable_balance: 1000000000000000000u128.into(), // 1 FLIP
			bound_redeem_address: None,
			restricted_balances: BTreeMap::new(),
			delegating_to: Some(AccountId32::new([0x33; 32])),
		},
		role_specific: RpcAccountInfo::LiquidityProvider {
			refund_addresses,
			earned_fees,
			boost_balances,
		},
	};

	insta::assert_snapshot!(serde_json::to_string_pretty(&wrapper).unwrap());
}

#[test]
fn test_validator_serialization() {
	let wrapper = RpcAccountInfoWrapper {
		common_items: RpcAccountInfoCommonItems {
			flip_balance: (FLIPPERINOS_PER_FLIP * 50).into(), // 50 FLIP
			asset_balances: any::AssetMap {
				eth: eth::AssetMap {
					eth: 3000000000000000000u128.into(), // 3 ETH
					flip: 0u128.into(),
					usdc: 2500000u128.into(), // 2.5 USDC
					usdt: 1500000u128.into(), // 1.5 USDT
				},
				btc: btc::AssetMap { btc: 75000000u128.into() }, // 0.75 BTC
				dot: dot::AssetMap { dot: 25000000000u128.into() }, // 2.5 DOT
				arb: arb::AssetMap { eth: 500000000000000000u128.into(), usdc: 1000000u128.into() },
				sol: sol::AssetMap { sol: 5000000000u128.into(), usdc: 750000u128.into() },
				hub: hub::AssetMap {
					dot: 10000000000u128.into(),
					usdc: 500000u128.into(),
					usdt: 250000u128.into(),
				},
			},
			bond: (FLIPPERINOS_PER_FLIP * 100).into(), // 100 FLIP
			estimated_redeemable_balance: (FLIPPERINOS_PER_FLIP * 25).into(), // 25 FLIP
			bound_redeem_address: Some(H160::from([0x44; 20])),
			restricted_balances: BTreeMap::from_iter(vec![
				(H160::from([0x55; 20]), (FLIPPERINOS_PER_FLIP * 10).into()), // 10 FLIP
				(H160::from([0x66; 20]), (FLIPPERINOS_PER_FLIP * 5).into()),  // 5 FLIP
			]),
			delegating_to: None,
		},
		role_specific: RpcAccountInfo::Validator {
			last_heartbeat: 150000,
			reputation_points: 850,
			keyholder_epochs: vec![100, 123, 124, 125],
			is_current_authority: true,
			is_bidding: false,
			is_current_backup: false,
			is_online: true,
			is_qualified: true,
			apy_bp: Some(375u32), // 3.75% APY
			operator: Some(AccountId32::new([0x77; 32])),
		},
	};

	insta::assert_snapshot!(serde_json::to_string_pretty(&wrapper).unwrap());
}
