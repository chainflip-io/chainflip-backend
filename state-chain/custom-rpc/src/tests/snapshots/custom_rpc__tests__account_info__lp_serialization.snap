---
source: state-chain/custom-rpc/src/tests/account_info.rs
expression: "serde_json::to_string_pretty(&wrapper).unwrap()"
---
{
  "flip_balance": "0x15af1d78b58c40000",
  "asset_balances": {
    "Ethereum": {
      "ETH": "0xffffffffffffffffffffffffffffffff",
      "FLIP": "0x7fffffffffffffffffffffffffffffff",
      "USDC": "0x989680",
      "USDT": "0x4c4b40"
    },
    "Polkadot": {
      "DOT": "0x174876e800"
    },
    "Bitcoin": {
      "BTC": "0x1dcd6500"
    },
    "Arbitrum": {
      "ETH": "0x1",
      "USDC": "0x2"
    },
    "Solana": {
      "SOL": "0x3",
      "USDC": "0x4"
    },
    "Assethub": {
      "DOT": "0xba43b7400",
      "USDT": "0x1e8480",
      "USDC": "0x2dc6c0"
    }
  },
  "bond": 0,
  "estimated_redeemable_balance": "0xde0b6b3a7640000",
  "role": "liquidity_provider",
  "refund_addresses": {
    "Ethereum": "0x1111111111111111111111111111111111111111",
    "Polkadot": "15hUeRCARvGw1Xjvuszo9ufT9bSqePGzwbjTeqZKxEeYq1i9",
    "Bitcoin": null,
    "Arbitrum": "0x2222222222222222222222222222222222222222",
    "Solana": "11111111111111111111111111111111"
  },
  "earned_fees": {
    "Ethereum": {
      "ETH": "0xde0b6b3a7640000",
      "FLIP": "0xffffffffffffffff",
      "USDC": "0x7ffffffffffffffe",
      "USDT": "0x3d090"
    },
    "Polkadot": {
      "DOT": "0x37e11d600"
    },
    "Bitcoin": {
      "BTC": "0x17d7840"
    },
    "Arbitrum": {
      "ETH": "0x1",
      "USDC": "0x2"
    },
    "Solana": {
      "SOL": "0x2",
      "USDC": "0x4"
    },
    "Assethub": {
      "DOT": "0x12a05f200",
      "USDT": "0x7a120",
      "USDC": "0xf4240"
    }
  },
  "boost_balances": {
    "Ethereum": {
      "ETH": [],
      "FLIP": [],
      "USDC": [],
      "USDT": []
    },
    "Polkadot": {
      "DOT": []
    },
    "Bitcoin": {
      "BTC": [
        {
          "fee_tier": 5,
          "total_balance": "0x5f5e100",
          "available_balance": "0x2faf080",
          "in_use_balance": "0x2faf080",
          "is_withdrawing": false
        },
        {
          "fee_tier": 10,
          "total_balance": "0xbebc200",
          "available_balance": "0x8f0d180",
          "in_use_balance": "0x2faf080",
          "is_withdrawing": true
        }
      ]
    },
    "Arbitrum": {
      "ETH": [],
      "USDC": []
    },
    "Solana": {
      "SOL": [],
      "USDC": []
    },
    "Assethub": {
      "DOT": [],
      "USDT": [],
      "USDC": []
    }
  }
}
